File: boj/b_2839.js
function solution(n) {
  if (n === 3) return 1;
  if (n === 4) return -1;
  const table = Array(n + 1).fill(-1);
  [table[3], table[5]] = [1, 1];
  for (let i = 6; i <= n; i += 1) {
    if (table[i - 3] === -1 && table[i - 5] === -1) table[i] = -1;
    else if (table[i - 3] === -1 || table[i - 5] === -1)
      table[i] = Math.max(table[i - 3], table[i - 5]) + 1;
    else table[i] = Math.min(table[i - 3], table[i - 5]) + 1;
  }
  return table[n];
}
console.log(solution(18));

File: boj/b_2839.py
import sys
sys.stdin = open("input.txt", "r") # 제거
input = sys.stdin.readline
def solution(n):
    if n == 3:
        return print(1)
    elif n == 4:
        return print(-1)
    table = [-1] * (n + 1)
    table[3], table[5] = 1, 1
    for i in range(6, n + 1):
        if table[i - 3] == -1 and table[i - 5] == -1:
            table[i] = -1
        elif table[i - 3] == -1 or table[i - 5] == -1:
            table[i] = max(table[i - 3], table[i - 5]) + 1
        else:
            table[i] = min(table[i - 3], table[i - 5]) + 1
    return print(table[n])
solution(int(input()))



